bundles/org.palladiosimulator.textual.tpcm/src/org/palladiosimulator/textual/tpcm/scoping/TPCMScopeProvider.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * generated by Xtext 2.21.0
+ */
+package org.palladiosimulator.textual.tpcm.scoping;
+
+import java.util.Collections;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EReference;
+import org.eclipse.xtext.EcoreUtil2;
+import org.eclipse.xtext.scoping.IScope;
+import org.eclipse.xtext.scoping.Scopes;
+import org.eclipse.xtext.scoping.impl.FilteringScope;
+import org.palladiosimulator.textual.tpcm.language.DomainInterfaceProvidedRole;
+import org.palladiosimulator.textual.tpcm.language.EntryLevelSystemCallAction;
+import org.palladiosimulator.textual.tpcm.language.Initialization;
+import org.palladiosimulator.textual.tpcm.language.Interface;
+import org.palladiosimulator.textual.tpcm.language.InterfaceRequiredRole;
+import org.palladiosimulator.textual.tpcm.language.InternalConfigurableInterface;
+import org.palladiosimulator.textual.tpcm.language.InternalInterfaceProvidedRole;
+import org.palladiosimulator.textual.tpcm.language.LanguagePackage;
+import org.palladiosimulator.textual.tpcm.language.LinkingResource;
+import org.palladiosimulator.textual.tpcm.language.Parameter;
+import org.palladiosimulator.textual.tpcm.language.ProcessingResource;
+import org.palladiosimulator.textual.tpcm.language.PropertyDefinition;
+import org.palladiosimulator.textual.tpcm.language.PropertyInitializer;
+import org.palladiosimulator.textual.tpcm.language.ResourceEntityType;
+import org.palladiosimulator.textual.tpcm.language.Role;
+import org.palladiosimulator.textual.tpcm.language.SEFF;
+import org.palladiosimulator.textual.tpcm.language.SEFFCallAction;
+import org.palladiosimulator.textual.tpcm.language.SEFFContent;
+import org.palladiosimulator.textual.tpcm.language.SEFFIterateAction;
+import org.palladiosimulator.textual.tpcm.language.Signature;
+
+/**
+ * This class contains custom scoping description.
+ * 
+ * See
+ * https://www.eclipse.org/Xtext/documentation/303_runtime_concepts.html#scoping
+ * on how and when to use it.
+ */
+public class TPCMScopeProvider extends AbstractTPCMScopeProvider {
+	@Override
+	public IScope getScope(EObject context, EReference reference) {
```

Comment:
this method is a bit lengthy perhaps we can restructure it somehow?!

Reply:
I split of a lengthy case into a separate method. Aside from that it is hard to sensibly shorten the method as it consists only of class checks of the context object and reference filtering. For each case, there is seldom more than a single statement. This is along Xtext's guidelines. 

Should the amount of cases grow extensively, we might need to think about some more elaborate mechanism. But for now, I'd vote for keeping it this way.